mne/encoding/feature.py

```diff
@@ -0,0 +1,341 @@
+"""A collection of classes for transforming, preprocessing, and fitting."""
+# Authors: Chris Holdgraf <choldgraf@gmail.com>
+#          Jona Sassenhagen <jona.sassenhagen@gmail.de>
+#          Jean-Remi King <jeanremi.king@gmail.com>
+#
+# License: BSD (3-clause)
+import numpy as np
+from scipy import sparse
+from ..utils import warn, _reject_data_segments
+
+
+class DataDelayer(object):
+    """Transformer for creating delayed versions of an input stimulus.
+
+    Parameters
+    ----------
+    time_window : array, shape (2,)
+        Specifies a minimum/maximum delay in seconds. In this case, all
+        delays between `min(time_window)` and `max(time_window)`
+        will be calculated using `sfreq`.
+    delays: array of floats, shape (n_delays,)
+        The time (in seconds) of each delay for specifying
+        pre-defined delays. Negative means timepoints in the past,
+        positive means timepoints in the future.
+    sfreq: float
+        The sampling frequency of the series. Defaults to 1.
+    """
+    def __init__(self, delays=None, time_window=None, sfreq=1.):
+        # Check if we need to creafte delays ourselves
+        if time_window is not None:
+            if sfreq is None:
+                raise ValueError('If time_window given, must give sfreq.')
+            if delays is not None:
+                raise ValueError('Delays must be None if time_window is given')
+            tmin, tmax = time_window
+            start_idx = int(round(tmin * sfreq))
+            delays = np.arange(start_idx, int(round(tmax * sfreq)) + 1) / sfreq
+        elif delays is None:
+            raise ValueError('One of `time_window` or `delays` must be given.')
+        self.delays = np.asarray(delays)
+        self.sfreq = float(sfreq)
+        self.feature_type = 'delay'
+
+    def fit(self, X, y=None):
+        """Create a time-lagged representation of X.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_times, n_features)
+            The input data to be time-lagged and returned.
+        """
+        X_del, names = _delay_timeseries(X, self.delays, sfreq=self.sfreq)
+        self.X_delayed_ = X_del
```

Comment:
I'm storing `X_` right now just because we haven't figured out a good way for keeping track of feature names / delays yet. So I've been using it to debug and make sure that things are behaving sensibly. I can remove once we agree on a way for making sure we know what features are what.